src/NHibernate/Engine/TypedValue.cs

```diff
@@ -17,15 +17,32 @@ public sealed class TypedValue
 		private readonly object value;
 		private readonly IEqualityComparer<TypedValue> comparer;
 
-		public TypedValue(IType type, object value)
+		/// <summary>
+		/// Constructor for typed value that may represent a simple value or a list value (for a parameter list).
+		/// If knowing what is value, use <see cref="TypedValue(IType, object, bool)"/> instead.
+		/// </summary>
+		/// <param name="type">The type of the value (or of its elements if it is a list value)</param>
+		/// <param name="value">The value.</param>
+		/// <remarks>The logic for infering if the value should be considered as a list value is minimal and will not
+		/// catch all cases, like hashset.</remarks>
+		public TypedValue(IType type, object value) : this (type, value, !type.IsCollectionType && value is ICollection && !type.ReturnedClass.IsArray)
 		{
+		}
+
+		/// <summary>
+		/// Construct a typed value.
+		/// </summary>
+		/// <param name="type">The type of the value (or of its elements if it is a list value)</param>
+		/// <param name="value">The value.</param>
+		/// <param name="isList"><see langword="true" /> if the value is a list value (for a parameter list),
+		/// <see langword="false" /> otherwise.</param>
+		public TypedValue(IType type, object value, bool isList)
+		{
+			if (isList && value != null && !(value is IEnumerable))
+				throw new ArgumentException($"{nameof(value)} must be an {nameof(IEnumerable)} when {nameof(isList)} is true", nameof(value));
```

Comment:
In case some mistake is done later, better check `value` is at least `IEnumerable` when `isList` is true. (The list comparator will consider as always equals values which are both not `IEnumerable`.)